### PR TITLE
Create options page and support non-proxied UDP option

### PR DESCRIPTION
**Pull request by guoweis**

### CODE REVIEW

File: src/content/extensions/multipleroutes/src/README.md [hunk 1]
```diff
@@ -1,15 +1,18 @@
 ## Chrome WebRTC Network Limiter
-Disables the WebRTC multiple-routes option in Chrome's privacy settings.
+Configures the WebRTC traffic routing options in Chrome's privacy settings.
 
 ★ What it does:
-This configures WebRTC to not use certain IP addresses:
+This configures WebRTC to not use certain IP addresses or protocol:
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/README.md]: or protocol => and protocols

File: src/content/extensions/multipleroutes/src/README.md [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
 - IP addresses not visible to the public internet (e.g. addresses like 192.168.1.2)
 - any public IP addresses associated with network interfaces that are not used for web traffic (e.g. an ISP-provided address, when browsing through a VPN)
+- Require WebRTC traffic to go through proxy servers as configured in Chrome. Since most of the proxy servers don't handle UDP, this effectively turns off UDP for now, until UDP proxy support is available in Chrome and such proxies are widely deployed.
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/README.md]: UDP for now, until => UDP until

File: src/content/extensions/multipleroutes/src/README.md [hunk 3]
```diff
@@ -1,15 +1,18 @@
...
  
 Once the extension is installed, WebRTC will only use public IP addresses associated with the interface used for web traffic, typically the same addresses that are already provided to sites in browser HTTP requests.
 
+However, this will not disable non-proxied UDP by default. It could be configured through the extension's Options page.
+
 ★ Notes:
-This extension may affect the performance of applications that use WebRTC for audio/video or real-time data communication. Because it limits the potential network paths, WebRTC may pick a path that results in significantly longer delay or lower quality (e.g. through a VPN). We are attempting to determine how common this is.
+This extension may affect the performance of applications that use WebRTC for audio/video or real-time data communication. Because it limits the potential network paths and protocol, WebRTC may pick a path which results in significantly longer delay or lower quality (e.g. through a VPN) or use TCP only through proxy servers which is not ideal for real-time communication. We are attempting to determine how common this is.
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/README.md]: and protocol => and protocols

File: src/content/extensions/multipleroutes/src/README.md [hunk 4]
```diff
@@ -1,15 +1,18 @@
...
 ★ What it does:
-This configures WebRTC to not use certain IP addresses:
+This configures WebRTC to not use certain IP addresses or protocol:
 - IP addresses not visible to the public internet (e.g. addresses like 192.168.1.2)
 - any public IP addresses associated with network interfaces that are not used for web traffic (e.g. an ISP-provided address, when browsing through a VPN)
+- Require WebRTC traffic to go through proxy servers as configured in Chrome. Since most of the proxy servers don't handle UDP, this effectively turns off UDP for now, until UDP proxy support is available in Chrome and such proxies are widely deployed.
  
 Once the extension is installed, WebRTC will only use public IP addresses associated with the interface used for web traffic, typically the same addresses that are already provided to sites in browser HTTP requests.
 
+However, this will not disable non-proxied UDP by default. It could be configured through the extension's Options page.
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/README.md]: Perhaps something like "The extension may also disable non-proxied UDP, but this is not on by default and must be configured using the extension's Options page."

File: src/content/extensions/multipleroutes/src/_locales/en/messages.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+   "NETLI_MULTIROUTES_CHECKBOX": {
+      "message": "Enables the WebRTC multiple-routes option in Chrome's privacy settings."
+   },
+   "NETLI_MULTIROUTES_DESCRIPTION": {
+      "message": "If enabled, Chrome will explore all possible routing options when using WebRTC to find the most performant path, possibly exposing user's private IP address. Otherwise, WebRTC traffic will be routed the same way as regular HTTP."
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/_locales/en/messages.json]: user's private IP address => private IP addresses

Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/_locales/en/messages.json]: "routed the same way as regular HTTP" => "will be routed through the default network route, similar to HTTP traffic".

File: src/content/extensions/multipleroutes/src/_locales/en/messages.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
+{
+   "NETLI_MULTIROUTES_CHECKBOX": {
+      "message": "Enables the WebRTC multiple-routes option in Chrome's privacy settings."
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/_locales/en/messages.json]: Might be more clear as "Enables the multiple routes when using WebRTC."

File: src/content/extensions/multipleroutes/src/_locales/en/messages.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+   },
+   "NETLI_MULTIROUTES_DESCRIPTION": {
+      "message": "If enabled, Chrome will explore all possible routing options when using WebRTC to find the most performant path, possibly exposing user's private IP address. Otherwise, WebRTC traffic will be routed the same way as regular HTTP."
+   },
+   "NETLI_NONPROXIEDUDP_CHECKBOX": {
+      "message": "Enables the WebRTC non-proxied UDP option in Chrome's privacy settings."
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/_locales/en/messages.json]: Might be more clear as "Enables in non-proxied UDP when using WebRTC."

File: src/content/extensions/multipleroutes/src/_locales/en/messages.json [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+   },
+   "NETLI_NONPROXIEDUDP_DESCRIPTION": {
+      "message": "If enabled, Chrome is allowed to use non-proxied UDP to connect to peers or TURN servers when using WebRTC. Since most proxy servers don't handle UDP, using UDP possibly exposes user's IP address. Turning this off effectively forces WebRTC to only use TCP for now, until UDP proxy support is available in Chrome and such proxies are widely deployed. As a result, it also might hurt media performance and increase the load for proxy servers."
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/_locales/en/messages.json]: user's IP address => IP addresses

Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/_locales/en/messages.json]: for now, until => until

File: src/content/extensions/multipleroutes/src/options.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
+// Saves options.
+function save_options() {
+  var multiRoutes = document.getElementById('multiRoutes').checked;
+  var nonProxiedUdp = document.getElementById('nonProxiedUdp').checked;
+  chrome.privacy.network.webRTCMultipleRoutesEnabled.set({
+    'value': multiRoutes
+  });
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/options.js]: Why is the first one in a try block but not the second?

File: src/content/extensions/multipleroutes/src/options.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+// Restores checkbox states.
+function restore_options() {
+  chrome.privacy.network.webRTCMultipleRoutesEnabled.get({},
+    function(details) {
+      document.getElementById('multiRoutes').checked =
+        details.value;
+    });
+  try {
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/options.js]: Same here: Only one is inside a try block.

File: src/content/extensions/multipleroutes/src/options.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    'value': multiRoutes
+  });
+  try {
+    chrome.privacy.network.webRTCNonProxiedUdpEnabled.set({
+      'value': nonProxiedUdp
+    });
+  } catch (err) {
+    document.getElementById('nonProxiedUdp').checked = false;
+    document.getElementById('nonProxiedUdp').disabled = true;
+  }
```
Comment (pthatcherg) [inline src/content/extensions/multipleroutes/src/options.js]: If something goes wrong, shouldn't we set it back to what is was?

document.getElementById('nonProxiedUdp').checked = nonProxiedUdp;

**Comment by KaptenJansson** [comment]
Why was this closed without a merge?

**Comment by guoweis-work** [comment]
This was replaced by pull request 648.

Thanks,
Guowei

**Pull request closed without merging**